tests/server/api/api_test.go

```diff
@@ -264,6 +264,126 @@ func (s *testMiddlewareSuite) TestAuditMiddleware(c *C) {
+	c.Assert(leader.GetServer().IsRateLimitMiddlewareEnabled(), Equals, true)
+
+	req, _ = http.NewRequest("GET", leader.GetAddr()+"/pd/api/v1/fail/", nil)
+	resp, err = dialClient.Do(req)
+	c.Assert(err, IsNil)
+	_, err = io.ReadAll(resp.Body)
+	resp.Body.Close()
+	c.Assert(err, IsNil)
+	c.Assert(resp.StatusCode, Equals, http.StatusTooManyRequests)
+
+	leader.SetRateLimiter(nil)
+	req, _ = http.NewRequest("GET", leader.GetAddr()+"/pd/api/v1/fail/", nil)
+	resp, err = dialClient.Do(req)
+	c.Assert(err, IsNil)
+	_, err = io.ReadAll(resp.Body)
+	resp.Body.Close()
+	c.Assert(err, IsNil)
+	c.Assert(resp.StatusCode, Equals, http.StatusOK)
+
+	req, _ = http.NewRequest("POST", leader.GetAddr()+"/pd/api/v1/admin/ratelimit-middleware?enable=false", nil)
+	resp, err = dialClient.Do(req)
+	c.Assert(err, IsNil)
+	resp.Body.Close()
+	c.Assert(leader.GetServer().IsRateLimitMiddlewareEnabled(), Equals, false)
+	c.Assert(failpoint.Disable("github.com/tikv/pd/pkg/ratelimiter/dontAllowRatelimiter"), IsNil)
+}
+
+func (s *testMiddlewareSuite) TestNoRequsetInfo(c *C) {
+	c.Assert(failpoint.Enable("github.com/tikv/pd/server/api/addAuditMiddleware", "return(true)"), IsNil)
+	c.Assert(failpoint.Enable("github.com/tikv/pd/server/api/addRateLimitMiddleware", "return(true)"), IsNil)
+	c.Assert(failpoint.Enable("github.com/tikv/pd/server/api/disableRequestInfoMiddleware", "return(true)"), IsNil)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	server.EnableZap = true
+	cluster, err := tests.NewTestCluster(ctx, 1)
+	c.Assert(err, IsNil)
+	c.Assert(cluster.RunInitialServers(), IsNil)
+	c.Assert(cluster.WaitLeader(), Not(HasLen), 0)
+	defer cancel()
+	defer cluster.Destroy()
+
+	leader := cluster.GetServer(cluster.GetLeader())
+
+	req, _ := http.NewRequest("POST", leader.GetAddr()+"/pd/api/v1/admin/audit-middleware?enable=true", nil)
+	resp, err := dialClient.Do(req)
+	c.Assert(err, IsNil)
+	resp.Body.Close()
+	c.Assert(leader.GetServer().IsAuditMiddlewareEnabled(), Equals, true)
+
+	req, _ = http.NewRequest("GET", leader.GetAddr()+"/pd/api/v1/fail/", nil)
+	resp, err = dialClient.Do(req)
+	c.Assert(err, IsNil)
+	_, err = io.ReadAll(resp.Body)
+	resp.Body.Close()
+	c.Assert(err, IsNil)
+	c.Assert(resp.StatusCode, Equals, http.StatusOK)
+	c.Assert(resp.Header.Get("audit-label"), Equals, "")
+
+	req, _ = http.NewRequest("POST", leader.GetAddr()+"/pd/api/v1/admin/ratelimit-middleware?enable=true", nil)
+	resp, err = dialClient.Do(req)
+	c.Assert(err, IsNil)
+	resp.Body.Close()
+	c.Assert(leader.GetServer().IsRateLimitMiddlewareEnabled(), Equals, true)
+
+	req, _ = http.NewRequest("GET", leader.GetAddr()+"/pd/api/v1/fail/", nil)
+	resp, err = dialClient.Do(req)
+	c.Assert(err, IsNil)
+	_, err = io.ReadAll(resp.Body)
+	resp.Body.Close()
+	c.Assert(err, IsNil)
+	c.Assert(resp.StatusCode, Equals, http.StatusOK)
+	c.Assert(resp.Header.Get("rate-limit"), Equals, "")
+
+	c.Assert(failpoint.Disable("github.com/tikv/pd/server/api/addAuditMiddleware"), IsNil)
```

Comment:
Is there any way to reduce the number of failpoints?

Reply:
It seems a good way that we can remove these failpoints if the audit backend has been implemented